### PR TITLE
CURATOR-621: Fix write acquire after downgrade InterProcessReadWriteLock

### PR DESCRIPTION
**Pull request by kezhuw**
Currently, in downgrading write lock of InterProcessReadWriteLock, read lock could have larger sorting sequence than contending write-acquire, this cause contending write-acquire to succeed after downgrading.

This commit solves this by using write lock's sorting sequence for read lock in downgrading.

### CODE REVIEW

**Comment by kezhuw** [comment]
@tisonkun Thank you for reviewing. Given the test case `testContendingDowngrading` as example.

1. Without this patch, `lock2` will succeed to acquire write after `lock1` downgrades its write to read since `lock2`'s write sequence is smaller than `lock1`'s read sequence.

```mermaid
sequenceDiagram

lock1 ->> Server: write acquire
Server ->> lock1: __WRITE__0000000000
Note right of lock1: write locked
lock2 -->> Server: write acquire
Server -->> lock2: __WRITE__0000000001
Note left of lock2: wait for __WRITE__0000000000
lock1 ->> Server: read acquire
Server ->> lock1: __READ__0000000002
activate lock1
Note right of lock1: read locked since:<br/> lock1 owns write lock
Note over lock1, lock2: lock1 holds write(0000000000) and read(0000000002), lock2 queues write(0000000001)
lock1 ->> Server: write release
par
Server ->> lock1: delete __WRITE__0000000000
Server -->> lock2: delete __WRITE__0000000000
end
activate lock2
Note left of lock2: write locked since:<br/>0000000001(lock2's write) < 0000000002(lock1's read)
Note over lock1,lock2:BUG: lock1 holds read(0000000002), lock2 holds write(0000000001)
deactivate lock2
deactivate lock1
```
2. With this patch applied, `lock1` will get same sorting sequence as its write lock in downgrading so there is no possibility that contending locks could get smaller sorting sequence.

**Review by tisonkun — APPROVED** [review]
+1

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessReadWriteLock.java
Comment (tisonkun) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessReadWriteLock.java:248]: nit: this test doesn't need two clients. You can refactor to use single client:

```java
        CuratorFramework client = CuratorFrameworkFactory.newClient(server.getConnectString(), new RetryOneTime(1));
```

**Review by eolivelli — APPROVED** [review]
+1

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessReadWriteLock.java
Comment (kezhuw) [reply curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessReadWriteLock.java]: Fixup commit added.

Comment (tisonkun) [reply curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessReadWriteLock.java]: @kezhuw I see. I'm waiting for further review the last days.

@Randgalt comments on the Slack workspace that he may review this patch the next week. Let's wait for a while. If there's no objection I'll merge it later the next week.

Comment (tisonkun) [reply curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessReadWriteLock.java]: @kezhuw as you're added in the PMC, I'm wondering if you're willing to make the next release (5.4.1 or 5.5.0) following [the instructions](https://cwiki.apache.org/confluence/display/CURATOR/For+Curator+Committers#ForCuratorCommitters-ReleasingCurator).

Comment (kezhuw) [reply curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessReadWriteLock.java]: @tisonkun I am willing to take care of next release. It is also a chance for me to familiar with the release procedure. I will start next release cycle once this pr solved and PMC permission granted(which might take up to one week according to secretary's reply).

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/locks/InterProcessReadWriteLock.java
Comment (Randgalt) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/locks/InterProcessReadWriteLock.java:116]: This old code doesn't do anything right?

**Review by Randgalt — APPROVED** [review]
It's been too long since I've looked at this code. I'll trust the test and explanation here.

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/locks/InterProcessReadWriteLock.java
Comment (kezhuw) [reply curator-recipes/src/main/java/org/apache/curator/framework/recipes/locks/InterProcessReadWriteLock.java:116]: Yep, I have same thought.

**Comment by tisonkun** [comment]
Merging...

**Comment by tisonkun** [comment]
@kezhuw Thanks for your contribution!

**Pull request merged**
